asynciterator.ts

```diff
@@ -1852,17 +1894,166 @@ class HistoryReader<T> {
+    this.close();
+  }
+
+  protected _onSourceError(err: Error) {
+    this.emit('error', err);
+  }
+
+  protected _end(destroy: boolean = false) {
+    super._end(destroy);
+    if (this._source) {
+      this._source.removeListener('end', this._onSourceEnd);
+      this._source.removeListener('error', this._onSourceError);
+      this._source.removeListener('readable', this._onSourceReadable);
+      delete this._source;
+    }
+  }
+
+  read(): T | null {
+    if (this._source)
+      return this._source.read();
+    return null;
+  }
+}
+
+export interface WrapOptions {
+  prioritizeIterable?: boolean;
+  letIteratorThrough?: boolean;
+}
+
+export type WrapSource<T> = AsyncIterator<T> | T[] | EventEmitter | Iterator<T> | Iterable<T>;
+
+function fromSource<T>(source: WrapSource<T> | undefined | null, options: WrapOptions = {}): AsyncIterator<T> {
+  if (!source)
+    return empty();
+  if (options.letIteratorThrough && source instanceof AsyncIterator)
+    return source;
+  if (options.prioritizeIterable && isIterableOrIterator<T>(source))
+    return fromIterableOrIterator<T>(source);
+  if (Array.isArray(source))
+    return fromArray<T>(source);
+  if (isAsyncIteratorLike<T>(source))
+    return fromAsyncIteratorLike<T>(source);
+  if (!options.prioritizeIterable && isIterableOrIterator<T>(source))
+    return fromIterableOrIterator<T>(source);
+  throw new Error(`Unsupported source ${source}`);
+}
+
 /**
   Creates an iterator that wraps around a given iterator or readable stream.
   Use this to convert an iterator-like object into a full-featured AsyncIterator.
   After this operation, only read the returned iterator instead of the given one.
   @function
-  @param {module:asynciterator.AsyncIterator|Readable} [source] The source this iterator generates items from
+  @param [sourceOrPromise] The source this iterator generates items from
   @param {object} [options] Settings of the iterator
   @returns {module:asynciterator.AsyncIterator} A new iterator with the items from the given iterator
 */
-export function wrap<T>(source: EventEmitter | Promise<EventEmitter>, options?: TransformIteratorOptions<T>) {
-  return new TransformIterator<T>(source as AsyncIterator<T> | Promise<AsyncIterator<T>>, options);
+export function wrap<T>(
+  sourceOrPromise: WrapSource<T> | Promise<WrapSource<T>> | null | undefined,
+  options: TransformIteratorOptions<T> & WrapOptions = {},
+): AsyncIterator<T> {
+  if ('maxBufferSize' in options || 'autoStart' in options || 'optional' in options || 'destroySource' in options)
+    return new TransformIterator<T>(sourceOrPromise as AsyncIterator<T> | Promise<AsyncIterator<T>>, options);
+  if (isPromise<T>(sourceOrPromise))
+    return new WrappingIterator(sourceOrPromise, options);
+  try {
+    return fromSource(sourceOrPromise as WrapSource<T>, options);
+  }
+  catch (err) {
+    const iterator = empty<T>();
+    scheduleTask(() => iterator.emit('error', err));
+    return iterator;
+  }
```

Comment:
This deviates from the way errors are specified in the readable spec https://nodejs.org/api/stream.html#event-error. Primarily, errors should only occur when *data* is being written or piped.

For the sake of future-proofing / compatibility with the `Readable` API - I would just throw an error instead rather than making it part of the stream.

Reply:
I had been under the impression that this would have constituted a breaking change but, upon looking at `main` again, I realized I was wrong. Done!